pkg/leader/leader.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leader
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ErrNoNS indicates that a namespace could not be found for the current
+// environment
+var ErrNoNS = errors.New("namespace not found for current environment")
+
+// maxBackoffInterval defines the maximum amount of time to wait between
+// attempts to become the leader.
+const maxBackoffInterval = time.Second * 16
+
+// TryBecome behaves like Become, except it will not return an error in the
+// case where a namespace cannot be found for the current pod. This is useful
+// for a service that might run outside the cluster, for example an operator
+// being started with `operator-sdk up local`.
+func TryBecome(ctx context.Context, lockName string) error {
+	err := Become(ctx, lockName)
+	if err == ErrNoNS {
+		logrus.Warn("leader election disabled; no lockNamespace was detected")
+		return nil
+	}
+	return err
+}
+
+// Become ensures that the current pod is the leader within its namespace. It
+// continuously tries to create a ConfigMap with the provided name and the
+// current pod set as the owner reference. Only one can exist at a time with
+// the same name, so the pod that successfully creates the ConfigMap is the
+// leader. Upon termination of that pod, the garbage collector will delete the
+// ConfigMap, enabling a different pod to become the leader.
+func Become(ctx context.Context, lockName string) error {
```

Comment:
How about a making the function name a little more descriptive?
`BecomeLeader()` or something like `GetLeaderLock()`.
Reading `Become()` isn't as obvious.

Reply:
Since it gets used as `leader.Become`, I think the namespace makes all clear. The third paragraph here describes my reasoning: https://golang.org/doc/effective_go.html#package-names

I'm also not crazy about starting the name with `Get`, because it doesn't return anything except an error.

Reply:
@hasbro17  I think `leader.Become` makes sense.

Reply:
Yeah my bad. I forgot that this would be called as `leader.Become()`. It's fine in that case.